ScriptBuffer IPC serialization is doing redundant work

kkinnunen-apple opened the pull request:
#### db14efcc24bcc45321da73cc77b502b0dcab3e66
<pre>
ScriptBuffer IPC serialization is doing redundant work
<a href="https://bugs.webkit.org/show_bug.cgi?id=303135">https://bugs.webkit.org/show_bug.cgi?id=303135</a>
<a href="https://rdar.apple.com/problem/165440987">rdar://problem/165440987</a>

Reviewed by Jean-Yves Avenard.

When serializing ScriptBuffer for IPC, serialization would create:
 - SharedMemory
 - ShareableResource
 - ShareableResourceHandle

ScriptBuffer contains a FragmentedSharedBuffer. We already have
IPC serialization for FragmentedSharedBuffer. We should simply use that.

Add FragmentedSharedBuffer serialization feature where single-segment
mmap-backed data is sent as directly shared memory.

This is work towards removing SharedMemory::wrapMap, which is unsafe
function and should be removed in favor of
SharedMemoryHandle::createVMShared() used here.

Removes
SharedBufferBuilder&amp; SharedBufferBuilder::operator=(RefPtr&lt;FragmentedSharedBuffer&gt;&amp;&amp;)
because it was buggy (did not reset()). It also is overly specific
operation for what it does.

Removes ShareBufferBuilder(RefPtr&lt;..SharedBuffer&gt;&amp;&amp;) constructor
variants, as they did not adopt the ptr, but implied that they did. Their
lack of const qualified ..SharedBuffer and use of rvalue RefPtr vs.
ShareBufferBuilder::append(const FragmentedSharedBuffer&amp;) created
inconsistent API surface.

* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::setData):
* Source/WebCore/bindings/js/ScriptBufferSourceProvider.h:
* Source/WebCore/bindings/js/WebAssemblyScriptBufferSourceProvider.h:
* Source/WebCore/loader/SubstituteResource.h:
(WebCore::SubstituteResource::SubstituteResource):
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::toIPCData const):
(WebCore::SharedBufferBuilder::initialize): Deleted.
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::SharedBufferBuilder::SharedBufferBuilder):
* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::ScriptBuffer):
(WebCore::tryConvertToShareableResourceHandle): Deleted.
(WebCore::ScriptBuffer::fromIPCData): Deleted.
(WebCore::ScriptBuffer::ipcData const): Deleted.
* Source/WebCore/workers/ScriptBuffer.h:
(WebCore::ScriptBuffer::buffer const):
(WebCore::ScriptBuffer::ScriptBuffer):
(WebCore::ScriptBuffer::protectedBuffer const): Deleted.
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::ensureCustomFontData):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::store):
(WebCore::SWScriptStorage::retrieve):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/303676@main">https://commits.webkit.org/303676@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c2ac58045c9cb4091f6266126e5c75147461bfd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133282 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5783 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44415 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140836 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a7df92d4-9a42-4b82-ad80-a118e72855da) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135152 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5648 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101955 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ed10a7f5-1a9e-43ef-be80-304b1eb12237) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136229 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119469 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82754 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d10319d4-3df9-4b71-b018-9b2cc6dabde9) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1936 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113437 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37584 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143484 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5453 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38161 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110331 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/5535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4688 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110516 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27999 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115723 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5508 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5354 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68960 "Built successfully") | | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5597 "Failed to checkout and rebase branch from PR 54481") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/loader/SubstituteResource.h:
```diff
@@ -48,7 +48,7 @@ class SubstituteResource : public RefCounted<SubstituteResource> {
         : m_url(WTFMove(url))
         , m_response(WTFMove(response))
-        , m_data(WTFMove(data))
+        , m_data(data)
```
why remove the move ?

jyavenard replied:
m_data is a SharedBufferBuilder I see. and you remove the move constructor.

jyavenard commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -587,29 +593,6 @@ bool DataSegment::containsMappedFileData() const
     return std::holds_alternative<FileSystem::MappedFileData>(m_immutableData);
 }
 
-SharedBufferBuilder::SharedBufferBuilder(RefPtr<FragmentedSharedBuffer>&& buffer)
```
this seems out of scope and IMHO deserves its own commit.

kkinnunen-apple replied:
Unfortunately the `RefPtr<FragmentedSharedBuffer>&&` constructor arg isn't workable because it needs to be `const FragmentedSharedBuffer`. Thus all constructors need to be changed. And while changing, the redundant Ref<>, and RefPtr<> and `..Builder::initialise` are removed here, as the only variant that is needed is the `const FragmentedSharedBuffer&`

jyavenard commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -672,6 +655,7 @@ void SharedBufferBuilder::appendDataSegment(Ref<DataSegment>&& segment)
 void SharedBufferBuilder::append(const FragmentedSharedBuffer& data)
 {
     m_state = State::Stale;
+    m_segments.reserveCapacity(m_segments.size() + data.segmentsCount());
```
unneeded, this is done by `appendContainerWithMapping` already

jyavenard commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -339,11 +339,7 @@ class SharedBufferBuilder {
-    explicit SharedBufferBuilder(RefPtr<SharedBuffer>&& buffer)
-        : SharedBufferBuilder(RefPtr<FragmentedSharedBuffer>{ WTFMove(buffer) }) { }
-    explicit SharedBufferBuilder(Ref<SharedBuffer>&& buffer) { initialize(WTFMove(buffer)); }
+    SharedBufferBuilder(const FragmentedSharedBuffer& buffer) { append(buffer); }
```
LGTM and it makes the code simpler, but we add a refcount churn where one wasn't needed before.

IIRC the previous version was done because a Ref<T>&& is directly convertible into a RefPtr<T>&& and compiler had issue differentiating the two.

kkinnunen-apple replied:
No ref churn added here. Nothing takes the ownership of Ref<..>&& or RefPtr<...>&& in the previous work.
Nothing takes a new ref in the new work.

jyavenard reviewed (approved):
LGTM with nits.

I would split this PR into two however. ShatedBuffer changes are unneeded for the core change

kkinnunen-apple committed:
ScriptBuffer IPC serialization is doing redundant work
https://bugs.webkit.org/show_bug.cgi?id=303135
rdar://problem/165440987

Reviewed by Jean-Yves Avenard.

When serializing ScriptBuffer for IPC, serialization would create:
 - SharedMemory
 - ShareableResource
 - ShareableResourceHandle

ScriptBuffer contains a FragmentedSharedBuffer. We already have
IPC serialization for FragmentedSharedBuffer. We should simply use that.

Add FragmentedSharedBuffer serialization feature where single-segment
mmap-backed data is sent as directly shared memory.

This is work towards removing SharedMemory::wrapMap, which is unsafe
function and should be removed in favor of
SharedMemoryHandle::createVMShared() used here.

Removes
SharedBufferBuilder& SharedBufferBuilder::operator=(RefPtr<FragmentedSharedBuffer>&&)
because it was buggy (did not reset()). It also is overly specific
operation for what it does.

Removes ShareBufferBuilder(RefPtr<..SharedBuffer>&&) constructor
variants, as they did not adopt the ptr, but implied that they did. Their
lack of const qualified ..SharedBuffer and use of rvalue RefPtr vs.
ShareBufferBuilder::append(const FragmentedSharedBuffer&) created
inconsistent API surface.

* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::FetchBodyConsumer::setData):
* Source/WebCore/bindings/js/ScriptBufferSourceProvider.h:
* Source/WebCore/bindings/js/WebAssemblyScriptBufferSourceProvider.h:
* Source/WebCore/loader/SubstituteResource.h:
(WebCore::SubstituteResource::SubstituteResource):
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::toIPCData const):
(WebCore::SharedBufferBuilder::initialize): Deleted.
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::SharedBufferBuilder::SharedBufferBuilder):
* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::ScriptBuffer):
(WebCore::tryConvertToShareableResourceHandle): Deleted.
(WebCore::ScriptBuffer::fromIPCData): Deleted.
(WebCore::ScriptBuffer::ipcData const): Deleted.
* Source/WebCore/workers/ScriptBuffer.h:
(WebCore::ScriptBuffer::buffer const):
(WebCore::ScriptBuffer::ScriptBuffer):
(WebCore::ScriptBuffer::protectedBuffer const): Deleted.
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::ensureCustomFontData):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::store):
(WebCore::SWScriptStorage::retrieve):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/303676@main
